R/ggMarginalHelpers.R

```diff
@@ -1,78 +1,139 @@
+  if (wasFlipped(scatPbuilt = scatPbuilt)) {
+    marg <- switch(marg,
+      "x" = "y",
+      "y" = "x"
+      )
   }
-  if (margin == "x") {
-    extraParams <- append(xparams, extraParams)
-    extraParams <- extraParams[!duplicated(names(extraParams))]
-    if (type == "histogram") {
-      if (!is.null(pb$layout$panel_scales$x[[1]]$get_limits)) {
-        extraParams[[originParamName]] <- pb$layout$panel_scales$x[[1]]$get_limits()[1]
-      }
-    }
-  } else if (margin == "y") {
-    extraParams <- append(yparams, extraParams)
-    extraParams <- extraParams[!duplicated(names(extraParams))]
-    if (type == "histogram") {
-      if (!is.null(pb$layout$panel_scales$y[[1]]$get_limits)) {
-        extraParams[[originParamName]] <- pb$layout$panel_scales$y[[1]]$get_limits()[1]
-      }
-    }      
+  var <- scatPbuilt[["data"]][[1]][[marg]]
+  
+  data.frame(var = var)
+}
+
+needsFlip <- function(marg, type) {
+  
+  # If the marginal plot is: (for the x margin (top) and is a boxplot) or 
+  #                          (for the y margin (right) and is not a boxplot), 
+  # ... then have to flip
+  topAndBoxP <- marg == "x" && type == "boxplot"
+  rightAndNonBoxP <- marg == "y" && type != "boxplot"
+  topAndBoxP || rightAndNonBoxP
+}
+
+margPlotNoGeom <- function(marg, type, data) {
+
+  # Build plot (sans geom). Note: Boxplot is the only plot type that needs y aes
+  if (type == "boxplot") {
+    plot <- ggplot2::ggplot(data = data, ggplot2::aes(x = var, y = var))
+  } else {
+    plot <- ggplot2::ggplot(data = data, ggplot2::aes(x = var))
+  }
+
+  if (needsFlip(marg = marg, type = type)) {
+    plot <- plot +  ggplot2::coord_flip()
   }
   
+  plot
+}
+
+alterParams <- function(marg, type, prmL, scatPbuilt) {
+
+  # merge the parameters in an order that ensures that marginal plot params 
+  # overwrite general params
+  prmL$exPrm <- append(prmL[[paste0(marg, "Prm")]], prmL$exPrm)
+  prmL$exPrm <- prmL$exPrm[!duplicated(names(prmL$exPrm))]
+
+  # pull out limit function and use if histogram
+  panScale <- getPanelScale(marg = marg, builtP = scatPbuilt)
+  lim_fun <- panScale$get_limits
+  if (type == "histogram" && !is.null(lim_fun)) {
+    prmL$exPrm[["boundary"]] <- lim_fun()[1]
+  } 
+  
+  # we're using geom_line for the density plot so that there will be no bottom line...
+  # ...so we have to tell geom_line to use non-default stat (stat = density)
   if (type == "density") {
-    extraParams[['stat']] <- "density"
-    layer <- do.call(ggplot2::geom_line, extraParams)
-  } else if (type == "histogram") {
-    layer <- do.call(ggplot2::geom_histogram, extraParams)
-  } else if (type == "boxplot") {
-    layer <- do.call(ggplot2::geom_boxplot, extraParams)
-  } else {
-    stop(sprintf("`type` = `%s` is not supported", type), call. = FALSE)
+    prmL$exPrm[['stat']] <- "density"
   }
+
+  prmL$exPrm
+}
+
+getPanelScale <- function(marg, builtP) {
+  panScale <- ifelse(utils::packageVersion("ggplot2") > "2.2.1", "_", "$")
```

Comment:
Here are the two possible versions:
```
# original
getPanelScale <- function(marg, builtP) {
  panScale <- ifelse(utils::packageVersion("ggplot2") > "2.2.1", "_", "$")
  txtExpr <- paste0("builtP$layout$panel_scales", panScale, marg, "[[1]]")
  panScaleExpr <- parse(text = txtExpr)
  eval(expr = panScaleExpr)
}

# no parse eval
getPanelScale <- function(marg, builtP) {
  above_221 <- utils::packageVersion("ggplot2") > "2.2.1"
  if (above_221) {
    if (marg == "x") {
      builtP$layout$panel_scales_x[[1]]
    } else {
      builtP$layout$panel_scales_y[[1]]
    }
  } else {
    if (marg == "x") {
      builtP$layout$panel_scales$x[[1]]
    } else {
      builtP$layout$panel_scales$y[[1]]
    }
  }
}
```
In my opinion the original version makes the logic clearer. It's not a big issue for me though, so just let me know which version we should go with.